Feat chat (#1362)
iotaledger/goshimmer: docs/tutorials/dApp.md

# How to create dApps

> This guide is meant for developers familiar with the Go programming language.

> **DISCLAIMER:** GoShimmer is a rapidly evolving prototype software. As such, the described steps here will likely change in the future. Specifically, we are envisioning to ease the process of dApp creation and installation for node owners. Furthermore, the current approach is in no way hardened and should be seen as purely experimental. Do not write any software for actual production use.

## Intro
Throughout this tutorial we will learn how to write simple dApps as GoShimmer plugins. We provide two different examples: a chat dApp and a network delay dApp. Hope you enjoy the reading!

## Chat dApp
In this guide we are going to explain how to write a very simple chat dApp so that anyone, connected to a GoShimmer node, could write a short message and read what is being written into the Tangle.

The complete source code of the application can be found [in the repository](https://github.com/iotaledger/goshimmer/tree/develop/plugins/chat). 

### Overview
Our chat dApp can be implemented in a few simple steps:

> 1. A node sends via the Tangle a special message containing a chat payload.

Reviewer: ```suggestion
1. A node sends a special message containing a chat payload via the Tangle.
```

> 2. Upon receipt, every other node in the network process this message and trigger an event that a chat message has beein received.

Reviewer: ```suggestion
2. Upon receipt, every other node in the network processes this message and - if the chat dApp/plugin is enabled - triggers an event that a chat message has been received.
```

> 2. Upon receipt, every other node in the network process this message and trigger an event that a chat message has beein received.

Reviewer 2: ```suggestion
2. Upon receipt, every other node in the network process this message and trigger an event that a chat message has been received.
```

Message<string>
```

Next, we need to fulfill the `Payload` interface and provide the functionality to read/write a payload from/to bytes. The [`hive.go/marshalutil`](https://github.com/iotaledger/hive.go/tree/master/marshalutil) package simplifies this step tremendously.
```Go
// Payload represents the generic interface for a payload that can be embedded in Messages of the Tangle.
type Payload interface {
    // Type returns the Type of the Payload.
    Type() Type
    
    // Bytes returns a marshaled version of the Payload.
    Bytes() []byte
    
    // String returns a human readable version of the Payload.
    String() string
}
```

Finally, we need to create and register our chat payload type so that it can be properly unmarshalled. 
```Go
// Type represents the identifier which addresses the chat payload type.
var Type = payload.NewType(payloadType, PayloadName, func(data []byte) (payload payload.Payload, err error) {
	var consumedBytes int
	payload, consumedBytes, err = FromBytes(data)
	if err != nil {
		return nil, err
	}
	if consumedBytes != len(data) {
		return nil, errors.New("not all payload bytes were consumed")
	}
	return
})
```

### Create The Web API Endpoints

> In order to issue a message with our newly created chat payload, we need to create a web API endpoint. Here we simply bind a json request containing the necessary fields: `from`, `to` and `message` and then issue it into the Tangle with `messagelayer.Tangle().IssuePayload(chatPayload)`. This plugin takes care of all the specifics and employs the `MessageFactory` to, i.a., select tips and sign the message.

Reviewer 2: ```suggestion
In order to issue a message with our newly created chat payload, we need to create a web API endpoint. Here we simply bind a json request containing the necessary fields: `from`, `to` and `message` and then issue it into the Tangle with `messagelayer.Tangle().IssuePayload(chatPayload)`. This plugin takes care of all the specifics and employs the `MessageFactory` to, i.e., select tips and sign the message.
```

Author: i.a., is correct in this context
